// @erebos/secp256k1/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.createKeyPair = createKeyPair;
exports.createPublic = createPublic;
exports.sign = sign;
exports.verify = verify;
exports.ec = void 0;

var _elliptic = _interopRequireDefault(require("elliptic"));

var ec = new _elliptic.default.ec('secp256k1');
exports.ec = ec;

function createKeyPair(privKey) {
  return privKey ? ec.keyFromPrivate(privKey, 'hex') : ec.genKeyPair();
}

function createPublic(pubKey) {
  return ec.keyFromPublic(pubKey, 'hex');
}

function sign(input, privKey) {
  var sig = ec.sign(input, privKey, {
    canonical: true
  });
  return [].concat(sig.r.toArray('be', 32), sig.s.toArray('be', 32), [sig.recoveryParam || 0]);
}

function verify(input, signature, pubKey) {
  var sig;

  if (Array.isArray(signature)) {
    var buf = Buffer.from(signature);
    sig = {
      r: buf.slice(0, 32),
      s: buf.slice(32, 64)
    };
  } else {
    sig = signature;
  }

  var key = typeof pubKey === 'string' ? createPublic(pubKey) : pubKey;
  return ec.verify(input, sig, key);
}
